### PR TITLE
tlf: Add DeepEqual verifying method.

### PR DESCRIPTION
**Pull request by jzila**
I need this to be able to check whether two handles are properly equal.

### CODE REVIEW

**Comment by songgao** [comment]
Is there a way to avoid using `reflect` here? We know all the fields and it seems we can just compare one-by-one.

**Comment by jzila** [comment]
Sigh, okaaaay. I was trying to avoid writing that much boilerplate.

**Comment by songgao** [comment]
Understood; sorry! I guess if we do this we should also add comment in the type definition so we don't forget updating the `DeepEqual` function if the type ever gets updated.

**Comment by jzila** [comment]
@songgao: good call. We have to do that anyway because of the sorts. And I could actually cache the sorting.

**Comment by strib** [comment]
Could add a static check like in `libkbfs.TlfHandle`: https://github.com/keybase/kbfs/blob/57d1e1d69b19b651ef466dd8761b1389bd51485e/libkbfs/tlf_handle.go#L271-L278

**Comment by jzila** [comment]
OK ready with feedback implemented.

File: tlf/handle.go [hunk 1]
```diff
@@ -77,6 +83,12 @@ func MakeHandle(
 	writers, readers []keybase1.UserOrTeamID,
 	unresolvedWriters, unresolvedReaders []keybase1.SocialAssertion,
 	extensions []HandleExtension) (Handle, error) {
+	if reflect.ValueOf(Handle{}).NumField() != 7 {
```
Comment (strib) [inline tlf/handle.go]: Can we do this in an `init()` instead, so it doesn't happen on every call?

File: tlf/handle.go [hunk 2]
```diff
@@ -326,3 +338,71 @@ func (h Handle) IsFinal() bool {
...
+	if len(h.UnresolvedReaders) != len(other.UnresolvedReaders) {
+		return false
+	}
+
+	if !h.sorted {
+		sort.Sort(UIDList(h.Writers))
+		sort.Sort(UIDList(h.Readers))
+		sort.Sort(SocialAssertionList(h.UnresolvedWriters))
+		sort.Sort(SocialAssertionList(h.UnresolvedReaders))
+		h.sorted = true
```
Comment (strib) [inline tlf/handle.go]: This doesn't have a pointer receiver, so isn't setting this caching field pretty useless?

File: tlf/handle.go [hunk 3]
```diff
@@ -326,3 +338,71 @@ func (h Handle) IsFinal() bool {
...
+	}
+	if !other.sorted {
+		sort.Sort(UIDList(other.Writers))
+		sort.Sort(UIDList(other.Readers))
+		sort.Sort(SocialAssertionList(other.UnresolvedWriters))
+		sort.Sort(SocialAssertionList(other.UnresolvedReaders))
+		other.sorted = true
```
Comment (strib) [inline tlf/handle.go]: Same here.

File: tlf/handle.go (hunk 2 above)
Comment (jzila) [reply tlf/handle.go]: Excellent point. Will fix.

File: tlf/handle.go [hunk 4]
```diff
@@ -326,3 +345,72 @@ func (h Handle) IsFinal() bool {
...
+		sort.Sort(UIDList(h.Readers))
+		sort.Sort(SocialAssertionList(h.UnresolvedWriters))
+		sort.Sort(SocialAssertionList(h.UnresolvedReaders))
+		h.sorted = true
+	}
+	if !other.sorted {
+		sort.Sort(UIDList(other.Writers))
+		sort.Sort(UIDList(other.Readers))
+		sort.Sort(SocialAssertionList(other.UnresolvedWriters))
+		sort.Sort(SocialAssertionList(other.UnresolvedReaders))
```
Comment (songgao) [inline tlf/handle.go]: Is it intended not to set `sorted` for `other`?

Comment (jzila) [reply tlf/handle.go]: Yeah, I want to be able to accept a non pointer comparer. Also, it seems wrong to mutate the comparison target.

**Pull request merged**